text/3437-implementable-trait-alias.md

```diff
@@ -0,0 +1,859 @@
+    type Item = i32; // Would be an error if this was `String`
+
+    fn next(&mut self) -> Self::Item {
+        42
+    }
+}
+
+struct Bar;
+impl !Send for Bar;
+
+//impl CopyIterator for Bar { /* ... */ } // ERROR: `Bar` is not `Send`
+```
+
+```rust
+trait Foo {}
+trait Bar = Foo where Self: Send;
+//impl<T> Bar for T {} // ERROR: Need to add `T: Send` bound
+```
+
+```rust
+#![feature(trivial_bounds)]
+trait Foo {}
+trait Bar = Foo where String: Copy;
+//impl Bar for () {} // ERROR: `String: Copy` not satisfied
+```
+
+Bounds on generic parameters are also enforced at the `impl` site.
+
+```rust
+trait Underlying<T> {}
+
+trait Alias<T: Send> = Underlying<T>;
+
+impl Alias<*const i32> for i32 {} // Error: `*const i32` is not `Send`
+```
+
+If the trait alias uniquely constrains a portion of the `impl` block, that part
+can be omitted.
+
+```rust
+pub trait IntIterator = Iterator<Item = i32> where Self: Send;
+
+struct Baz;
+
+impl IntIterator for Baz {
+    // The alias constrains `Self::Item` to `i32`, so we don't need to specify it
+    // (though we are allowed to do so if desired).
+
+    fn next(&mut self) -> i32 {
+        -27
+    }
+}
+```
+
+Such constraints can be inferred indirectly:
+
+```rust
+trait Bar: Iterator<Item = i32> {}
+pub trait IntIterator = Iterator where Self: Bar;
+
+struct Baz;
+
+impl Bar for Baz {}
+
+impl IntIterator for Baz {
+    // `IntIterator` requires `Bar`,
+    // which requires `Iterator<Item = i32>`,
+    // so `Item` must be `i32`
+    // and we don't need to specify it.
+
+    fn next(&mut self) -> i32 {
+        -27
+    }
+}
+```
+
+Alias `impl`s also allow omitting implied `#[refine]`s:
+
+```rust
+//! crate frob-lib
+#![feature(trait_alias)]
+
+pub trait LocalFrob {
+    async fn frob(&self);
+}
+
+// not `+ Send`!
+pub trait Frob = LocalFrob<frob(..): Send> where Self: Send;
+```
+
+```rust
+//! crate joes-crate
+#![feature(trait_alias_impl)]
+
+use frob_lib::Frob;
+
+struct MyType;
+
+impl Frob for MyType {
+    // The return future of this method is implicitly `Send`, as implied by the alias.
+    // No `#[refine]` is necessary.
+    async fn frob(&self) { /* ... */ }
+}
+```
+
+Trait aliases are `unsafe` to implement iff one or more primary traits are
+marked `unsafe`.
+
+## Usage in paths
+
+Trait aliases can also be used with trait-qualified and fully-qualified method
+call syntax, as well as in paths more generally. When used this way, they are
+treated equivalently to the underlying primary trait(s), with the additional
+restriction that all `where` clauses and type parameter/associated type bounds
+must be satisfied.
+
+```rust
+use std::array;
+
+trait IntIter = Iterator<Item = u32> where Self: Clone;
+
+let iter = [1_u32].into_iter();
+let _: IntIter::Item = IntIter::next(&mut iter); // works
+let _: <array::IntoIter as IntIter>::Item = <array::IntoIter as IntIter>::next(); // works
+//IntIter::clone(&iter); // ERROR: trait `Iterator` has no method named `clone()`
```

Comment:
right below you state items from traits in `where` clauses are accessible, so `IntIter::clone` should be accessible